### PR TITLE
Add ObjC route_guide sample

### PR DESCRIPTION
**Pull request by jcanizales**
I would think the easiest way to review is to clone the branch, `pod install` in the sample directory, `make; ./route_guide_server` in the C++ directory, and open and run the App.

### CODE REVIEW

**Comment by murgatroid99** [comment]
I tried to run it, but the RouteGuideClient project only has the option to run it on an iOS device, and not a simulator, so I can't figure out how to run it.

**Comment by jcanizales** [comment]
That's super weird, it has the options alright on my machine. Let's look at it once I get there.

**Comment by murgatroid99** [comment]
LGTM if you add the documentation.

**Comment by jcanizales** [comment]
Done; PTAL.

File: objective-c/route_guide/README.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+Our example is a simple route mapping application that lets clients get information about features on their route, create a summary of their route, and exchange route information such as traffic updates with the server and other clients.
+
+You also should have [Cocoapods](https://cocoapods.org/#install) installed, as well as the relevant tools to generate the client library code (and a server in another language, for testing). You can obtain the latter by following [these setup instructions](https://github.com/grpc/homebrew-grpc).
+
+
+<a name="try"></a>
+## Try it out!
+
+To try the sample app, we need a gRPC server running locally. Let's compile and run, for example, the C++ server in this repository:
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: I'm not *entirely* convinced by having this section here (running the server and client and showing the output) rather than at the end, as in all the other tutorials, though that could be just my own personal preference.

It seems like an approach more suited to a Quick Start/demo (which we have for all the other languages and will need for Objective C too) - do we have an Objective-C greeter/hello service too?

File: objective-c/route_guide/README.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+$ pod install
+```
+
+which, before installing the generated library in the XCode project of this sample, runs:
+
+```shell
+$ protoc -I ../../protos --objc_out=Pods/RouteGuide --objcgrpc_out=Pods/RouteGuide ../../protos/route_guide.proto
+```
+
+Running this command generates the following files in under `Pods/RouteGuide/`:
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: in under -> in (or under, but not both :) )

File: objective-c/route_guide/README.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+option objc_class_prefix = "RTG";
+```
+
+
+<a name="protoc"></a>
+## Generating client code
+
+Next we need to generate the gRPC client interfaces from our .proto service definition. We do this using the protocol buffer compiler (`protoc`) with a special gRPC Objective-C plugin.
+
+For simplicity, we've provided a [Podspec file](https://github.com/grpc/grpc-common/blob/master/objective-c/route_guide/RouteGuide.podspec) that runs `protoc` for you with the appropriate plugin, input, and output, and describes how to compile the generated files. You just need to run in this directory:
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: "this directory" - which directory?`grpc-common/objective-c/route_guide`? (I know we've told people to change to this dir but we did so about two pages ago...)

File: objective-c/route_guide/README.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+(This might have to compile OpenSSL, which takes around 15 minutes if Cocoapods doesn't have it yet on your computer's cache).
+
+Finally, open the XCode workspace created by Cocoapods, and run the app. You can check the calling code in `ViewControllers.m` and see the results in XCode's log console.
+
+The next sections guide you step-by-step through the creation of this proto service, the client library generated from it, and the app that uses that library.
+
+
+<a name="proto"></a>
+## Defining the service
+
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: We need to rephrase this section a bit, as unlike the other tutorials our example is using an existing server/service so the user doesn't need to define it, just use it.

File: objective-c/route_guide/README.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+```shell
+$ pod install
+```
+
+(This might have to compile OpenSSL, which takes around 15 minutes if Cocoapods doesn't have it yet on your computer's cache).
+
+Finally, open the XCode workspace created by Cocoapods, and run the app. You can check the calling code in `ViewControllers.m` and see the results in XCode's log console.
+
+The next sections guide you step-by-step through the creation of this proto service, the client library generated from it, and the app that uses that library.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: See comment below, we're not really creating the service (as it's already being used in the server and we're not creating our own server) - how about:
guide you step-by-step through how this proto service is defined, how to generate a client library from it, and how to create an app that uses that library.

File: objective-c/route_guide/README.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+
+<a name="proto"></a>
+## Defining the service
+
+Our first step is to define the gRPC *service* and the method *request* and *response* types using [protocol buffers](https://developers.google.com/protocol-buffers/docs/overview). You can see the complete .proto file in [`grpc-common/protos/route_guide.proto`](https://github.com/grpc/grpc-common/blob/master/protos/route_guide.proto).
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: First let's look at how the service we're using is defined. A gRPC *service* and its method *request* and *response* types are defined using [protocol buffers](https://developers.google.com/protocol-buffers/docs/overview). You can see the complete .proto file for our example in [`grpc-common/protos/route_guide.proto`](https://github.com/grpc/grpc-common/blob/master/protos/route_guide.proto).

File: objective-c/route_guide/README.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+- `RouteGuide.pbobjc.h`, the header which declares your generated message classes.
+- `RouteGuide.pbobjc.m`, which contains the implementation of your message classes.
+- `RouteGuide.pbrpc.h`, the header which declares your generated service classes.
+- `RouteGuide.pbrpc.m`, which contains the implementation of your service classes.
+
+These contain:
+- All the protocol buffer code to populate, serialize, and retrieve our request and response message types.
+- A class called `RTGRouteGuide` that for clients to call with the methods defined in the `RouteGuide` service.
+
+The provided Podspec file works for any proto library you define; you just need to replace the name (matching the file name), version, and other metadata.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: Slight rephrase:

You can also use the provided Podspec file to generate client code from any other proto service definition; just replace the name (matching the file name), version, and other metadata.

File: objective-c/route_guide/README.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+Running this command generates the following files in under `Pods/RouteGuide/`:
+- `RouteGuide.pbobjc.h`, the header which declares your generated message classes.
+- `RouteGuide.pbobjc.m`, which contains the implementation of your message classes.
+- `RouteGuide.pbrpc.h`, the header which declares your generated service classes.
+- `RouteGuide.pbrpc.m`, which contains the implementation of your service classes.
+
+These contain:
+- All the protocol buffer code to populate, serialize, and retrieve our request and response message types.
+- A class called `RTGRouteGuide` that for clients to call with the methods defined in the `RouteGuide` service.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: "that for clients to call": that what for clients to call?  Maybe rephrase to 

 A class called `RTGRouteGuide` that lets clients call the methods defined in the `RouteGuide` service.

File: objective-c/route_guide/README.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+Notice that we've specified the HTTP scheme in the host address. This is because the server we will be using to test our client doesn't use [TLS](http://en.wikipedia.org/wiki/Transport_Layer_Security). This is fine because it will be running locally on our development machine. The most common case, though, is connecting with a gRPC server on the internet, running gRPC over TLS. For that case, the HTTPS scheme can be specified (or no scheme at all, as HTTPS is the default value). The default value of the port is that of the scheme selected: 443 for HTTPS and 80 for HTTP.
+
+
+### Calling service methods
+
+Now let's look at how we call our service methods. As you will see, all these methods are asynchronous, so you can call them from the main thread of your app without worrying about freezing your UI or the OS killing your app.
+
+#### Simple RPC
+
+Calling the simple RPC `GetFeature` is nearly as straightforward as calling any other aynschronous method on Cocoa.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: Nit: aynschronous -> asynchronous

File: objective-c/route_guide/README.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+[client getFeatureWithRequest:point handler:^(RTGFeature *response, NSError *error) {
+  if (response) {
+    // Successful response received
+  } else {
+    // RPC error
+  }
+}];
+```
+
+As you can see, we create and populate a request protocol buffer object (in our case `RTGPoint`). Then, we call the method on the client object, passing it the request, and a block to handle the response (or any RPC error). If the RPC finished successfully, the handler block is called with a `nil` error argument, and we can read the response information from the server from the response argument. If, instead, some RPC error happened, the handler block is called with a `nil` response argument, and we can read the details of the problem from the error argument.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: finished successfully -> finishes successfully
error happened -> error happens

File: objective-c/route_guide/README.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+  } else if (error) {
+    // RPC error; the stream is over.
+  }
+  if (done) {
+    // The stream is over (all the responses were received, or an error occured). Do any cleanup.
+  }
+}];
+```
+
+Notice how the signature of the handler block now includes a `BOOL done` parameter. The block will be called any number of times; only on the last call will the `done` argument be `YES`. If an error occurs, the RPC will be finished and the handler will be called with arguments `(YES, nil, error)`.
```
Comment (LisaFC) [inline objective-c/route_guide/README.md]: The handler block can be called any number of times; only on the last call is the `done` argument value set to `YES`. If an error occurs, the RPC finishes and the handler is called with the arguments `(YES, nil, error)`.

(I know what you're getting at, but in general I like to keep things in a sort of continuous present. Otherwise you can end up in a situation of peculiarly wandering tenses, where one minute you're in the past, then in the future, then conditional, etc. etc.)

**Comment by LisaFC** [comment]
Tutorial text looks really good overall. I'm wondering if/should we can have some sort of Objective-C equivalent of the Hello World/Quick Start as well? Obviously it won't be quite the same as it'll just be a client, we might have to point users to one of the other languages to get a running server.

File: objective-c/route_guide/README.md (hunk 4 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Well, if the user isn't defining the service herself, then there isn't much point about her following a generic tutorial at all: She'd just tell Cocoapods the app depends on someone's published generated library.

File: objective-c/route_guide/README.md (hunk 5 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Yeah I understand using "service" as protocol buffers' syntax uses it is confusing. In the real world, service and server are almost synonyms. But you define protocol buffer APIs using the keyword `service`. We tried to have that changed to `interface` for proto3, without success. For this tutorial I'm assuming the user defines the API (otherwise .proto files don't enter the picture).

File: objective-c/route_guide/README.md (hunk 4 above)
Comment (LisaFC) [inline objective-c/route_guide/README.md]: True, but OTOH at no point are we telling her in this tutorial how to create a server - you're unlikely to define a service and then just write clients for it even though it doesn't have any servers. Unless it's a reasonably likely scenario that an Objective C user would define a service, get someone else to implement the server in another language, then create a client app?

That's why I suggested rephrasing it as "let's look at how this service is defined" rather than "first you need to define this service", as the service is clearly already defined and implemented on the server side (and indeed running) in our scenario. It still provides enough info for the user to create their own service if they so wish.

Comment (LisaFC) [inline objective-c/route_guide/README.md]: I do think it only requires the small tweaks I suggested to be consistent, though.

File: objective-c/route_guide/README.md (hunk 11 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 10 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 9 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 8 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 7 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 6 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

File: objective-c/route_guide/README.md (hunk 3 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Clarified.

File: objective-c/route_guide/README.md (hunk 2 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: Done.

**Comment by jcanizales** [comment]
Yes, we want to have a Hello World and also an OAuth2 sample in place. Michael and I'll be writing them this week.

File: objective-c/route_guide/README.md (hunk 4 above)
Comment (jcanizales) [inline objective-c/route_guide/README.md]: > Unless it's a reasonably likely scenario that an Objective C user would define a service, get someone else to implement the server in another language, then create a client app?

What I had in mind was the same user would write the server, but not in Objective-C :)

Yeah, your wording works for all three scenarios, so I'll just change to that.

**Comment by jcanizales** [comment]
I think your comment on the ordering of sections got swallowed by GitHub. I changed it from the others because it's easy to jump over, and some people just want to look at and play with the code in their IDE. When I had looked earlier at the other languages for a server to run, I'd almost missed the "Try it out!" section, as it was buried below a lot of prose and code snippets.

I think with the table of contents in place it's easier to discover anyway, so I wouldn't mind too much moving it back to the bottom.

**Comment by LisaFC** [comment]
The ordering actually doesn't bother me that much (which is why I deleted the comment after reading through the rest of the text) - especially as the structure is already different from the other tutorials as it doesn't cover the server side.

**Comment by LisaFC** [comment]
Docs LGTM.
